### PR TITLE
New liquidity mining flow

### PR DESCRIPTION
**Pull request by cuong-incognito**
- Liquidity mining takes from DAO
- Lock share amount for the liquidity mining reward

### CODE REVIEW

File: blockchain/pdex/state_v2.go [hunk 1]
```diff
@@ -120,6 +120,50 @@ func (s *stateV2) Process(env StateEnvironment) error {
 		}
 	}
 
+	// unlock LM locked share amount
```
Comment (dat-incognito-org) [inline blockchain/pdex/state_v2.go:123]: this should be its own function

File: blockchain/pdex/state_v2.go [hunk 2]
```diff
@@ -307,6 +353,50 @@ func (s *stateV2) BuildInstructions(env StateEnvironment) ([][]string, error) {
...
+	for _, poolPair := range s.poolPairs {
```
Comment (dat-incognito-org) [inline blockchain/pdex/state_v2.go:357]: this should be its own function

File: blockchain/pdex/v2utils/state_change.go [hunk 3]
```diff
@@ -147,3 +153,43 @@ func GetChangedElementsFromMapStringBigInt(map0, map1 map[string]*big.Int) map[s
 	}
 	return res
 }
+
+func GetChangedElementsFromMapStringMapUint64(map0, map1 map[string]map[uint64]uint64) map[string]map[uint64]bool {
```
Comment (dat-incognito-org) [inline blockchain/pdex/v2utils/state_change.go:157]: with this many types it gets very difficult to review / distinguish these helpers. There should be a cleaner way. One solution might be: declaring a type, then define the function `GetChangedElements()` taking that type as receiver.
e.g `type DiffMapStringUint64Uint64 map[string]map[uint64]uint64`, `func (dm DiffMapStringUint64Uint64) GetChangedElements(other DiffMapStringUint64Uint64) DiffMapStringUint64Uint64`

File: blockchain/pdex/state_v2.go (hunk 1 above)
Comment (cuong-incognito) [reply blockchain/pdex/state_v2.go]: resolved

File: blockchain/pdex/v2utils/state_change.go (hunk 3 above)
Comment (tin-incognito) [reply blockchain/pdex/v2utils/state_change.go:157]: ok might be better than current

Comment (tin-incognito) [reply blockchain/pdex/v2utils/state_change.go:157]: updated

**Review by tin-incognito — APPROVED** [review]
LGTM!

**Review by dat-incognito-org — APPROVED** [review]

**Pull request merged**